pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java

```diff
@@ -121,13 +124,21 @@ private void assignWorkersToNonRootFragment(PlanFragment fragment, DispatchableP
     }
   }
 
-  private boolean isLocalExchange(List<PlanFragment> children) {
-    if (children.size() != 1) {
-      return false;
+  /**
+   * Returns the index of the child fragment that has a local exchange (SINGLETON distribution), or {@code null} if none
+   * exists.
+   */
+  @Nullable
+  private Integer findLocalExchange(List<PlanFragment> children) {
+    int numChildren = children.size();
+    for (int i = 0; i < numChildren; i++) {
+      PlanNode childPlanNode = children.get(i).getFragmentRoot();
+      if (childPlanNode instanceof MailboxSendNode
+          && ((MailboxSendNode) childPlanNode).getDistributionType() == RelDistribution.Type.SINGLETON) {
+        return i;
+      }
     }
-    PlanNode childPlanNode = children.get(0).getFragmentRoot();
-    return childPlanNode instanceof MailboxSendNode
-        && ((MailboxSendNode) childPlanNode).getDistributionType() == RelDistribution.Type.SINGLETON;
```

Comment:
nit: why not return -1 and avoid allocating object?

Reply:
Sure